taskcluster/translations_taskgraph/transforms/find_upstreams.py

```diff
@@ -0,0 +1,136 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# This transform sequence sets `dependencies` and `fetches` based on
+# the information provided in the `upstreams-config` data in each job
+# and the given parameters.
+
+# It will through all tasks generated from `kind-dependencies` and
+# set any tasks that match the following conditions as dependencies:
+# - src and trg locale given match the {src,trg}_locale attributes on the upstream task
+# - `upstream-task-attributes` given match their equivalents on the upstream task
+# - `dataset` attribute on the upstream task is one of the datasets provided in `parameters`
+#   for the `dataset-category` given in the job.
+#
+# Additionally, fetches will be added for those tasks for each entry in `upstream-artifacts`.
+#
+# (It is not ideal that this transform hardcodes dataset handling, but because kinds are
+# completely unaware of parameters, there's no other real way to do this.)
+
+import copy
+
+from taskgraph.transforms.base import TransformSequence
+from taskgraph.util.schema import Schema, optionally_keyed_by, resolve_keyed_by
+from voluptuous import ALLOW_EXTRA, Required
+
+SCHEMA = Schema(
+    {
+        Required("upstreams-config"): {
+            Required("locale-pair"): {
+                Required("src"): str,
+                Required("trg"): str,
+            },
+            Required("upstream-task-attributes"): {
+                str: optionally_keyed_by("cleaning-type", str),
+            },
+            Required("upstream-artifacts"): [str],
+        },
+    },
+    extra=ALLOW_EXTRA,
+)
+
+by_locales = TransformSequence()
+by_locales.add_validate(SCHEMA)
+
+def get_cleaning_type(src, trg, upstreams):
+    candidates = set()
+
+    for upstream in upstreams:
+        if upstream.kind not in ("bicleaner", "clean"):
+            continue
+
+        if upstream.attributes["src_locale"] != src or upstream.attributes["trg_locale"] != trg:
+            continue
+
+        candidates.add(upstream.attributes["cleaning-type"])
+
+    for type_ in ("bicleaner-ai", "bicleaner", "clean"):
+        if type_ in candidates:
+            return type_
+
+    raise Exception(f"Unable to find cleaning type for {src_locale}-{trg_locale}!")
```

Comment:
These `src_locale` and `trg_locale` variables are not defined.
```suggestion
    raise Exception(f"Unable to find cleaning type for {upstream.attributes['src_locale']}-{upstream.attributes['trg_locale']}!")
```